Filter by issue tags on board and history

locmai1 opened the pull request:
# Description

Previously, we filtered by keys of tags that we used in the join queue form, now we want to filter by topics which is a part of the Issue key in course's tags field:

- Removed unnecessary sorting topics since I feel like it's not needed? (let me know tho, can revert)
- Added a new util function to get the topic tags for modularity

Other things addressed:

- Fixed an infinite loading + firebase invalid query bug on `board/history` when users load the history board
- Restructure `cs160` tags map in firebase to match `cs40` tags (Issue, General, How do you feel about the topic)

## Issues

#74 

## Screenshots

Board + History before:

<img width="385" alt="image" src="https://github.com/user-attachments/assets/c5f4ad6a-797f-4804-80d5-b6a8f15b5950" />

<img width="385" alt="image" src="https://github.com/user-attachments/assets/c746554d-e76b-4beb-b398-98ef57cc8bcd" />

Board + History after:

<img width="384" alt="image" src="https://github.com/user-attachments/assets/cc29e493-784a-4473-975a-a9ea87533807" />

<img width="383" alt="image" src="https://github.com/user-attachments/assets/ceff1c59-1501-4fa2-863d-e71979b22804" />

## Test

Tested current questions on Board + expiredQuestions on History for both CS160 and CS40

## Possible Downsides

None :)

## Additional Documentations

None :)

locmai1 committed:
fix: patched infinite loading on history

locmai1 committed:
feat: issue filter on board

johnny-t06 commented on src/app/utils/index.ts:
```diff
@@ -209,3 +209,7 @@ export const getEmailTemplate = (type: string, email: string) => {
+
+export const getCourseTopicTags = (tags: Tags, tagKey: string): string[] => {
+  return (tags[tagKey].options || []).map((option) => option.choice);
+};
```
Nice!

nickbar01234 replied:
nit: Prefer `??` over `||`, since `tags[tagKey].options` is undefined. See [thread](https://stackoverflow.com/a/61481631/19129136) for a discussion on when to use what

johnny-t06 replied:
^^

johnny-t06 reviewed (approved):
LGTM, can wait for discussions about the sorting

locmai1 committed:
Merge master

locmai1 committed:
feat: null check for tags

locmai1 committed:
feat: re-style form + form ordering

locmai1 commented:
# Description

Previously, tags were ordered lexicographically. This PR also addresses the order of tags in both the join queue form as well as on the board, board/history, and displaying "How do you feel about the topic" only for TA view. Addressed:

- Tags are now ordered: General, Issue, How do you feel about the topic

Additional:

- Changed the text primary and secondary to reflect Figma
- Re-styled form to match Figma 
- Fixed the white border on tag selection in form

## Screenshots

Join queue form (before, after):

<img width="348" alt="image" src="https://github.com/user-attachments/assets/ea9209a9-9e8c-4139-9239-72ea7c6a53f1" />

<img width="348" alt="image" src="https://github.com/user-attachments/assets/1fb78cf3-e0de-413d-ac40-d397f33ac569" />


Board (user, TA):

<img width="354" alt="image" src="https://github.com/user-attachments/assets/0a78096a-d9cd-4755-a5ee-ea7425a21126" />

<img width="351" alt="image" src="https://github.com/user-attachments/assets/08977814-58ac-4036-bb86-1d4d7061bcb6" />

Board history (user, TA):

<img width="349" alt="image" src="https://github.com/user-attachments/assets/c00a7a9d-2fa4-4556-9121-94c6c375d5f9" />

<img width="346" alt="image" src="https://github.com/user-attachments/assets/544df463-b50c-406c-8560-db19b1f70ccb" />



## Test

All tests are done in photos above

## Possible Downsides

- Discovered a bug: TAs are join public board queues if they click on a question on the public board
- Right now, the TAG_ORDER is a constant but not sure if it needs to be dynamic or anything in future
- Joining queue form selection's label turn blue on filling form area but I think that's just MUI FormLabel

johnny-t06 commented on src/app/components/queue/form/QuestionForm.tsx:
```diff
@@ -5,10 +5,13 @@ import {
+  Typography,
 } from "@mui/material";
 import CloseIcon from "@mui/icons-material/Close";
+// import FileUploadOutlinedIcon from "@mui/icons-material/FileUploadOutlined";
```
nit: can remove

johnny-t06 reviewed (approved):
LGTM

locmai1 committed:
Merge branch 'master' into loc/filter-issue-tags

locmai1 committed:
cleanup: remove unused

locmai1 committed:
cleanup: simplify topic fn
